### PR TITLE
Added plugin system, added @git/fetch, @install/node, echo and @git/gitignore

### PR DESCRIPTION
**Pull request by FishingHacks**

### CODE REVIEW

**Comment by FishingHacks** [comment]
#2 - ignore that, I thought it was on FishingHacks/redstart ^^

**Comment by RedCrafter07** [comment]
You may rename some variables, as they're prefixed with "_" which indicates they are not being used. Thanks :)

**Comment by RedCrafter07** [comment]
Also, have you pulled the current version before creating this request?

**Comment by RedCrafter07** [comment]
If you'd include timestamps for every module, it would be great. Thank you 👍

**Comment by FishingHacks** [comment]
done

**Comment by RedCrafter07** [comment]
I can't see the timestamps anywhere 🤔

**Comment by FishingHacks** [comment]
I removed them, cuz i think they're unnecessary until I've implemented top-level config

**Comment by RedCrafter07** [comment]
I think you should see how long a plugin took to do its work

**Comment by RedCrafter07** [comment]
But I can implement it myself then. Thanks for these great ideas 👍 😃

**Comment by RedCrafter07** [comment]
I'll look it over again in a few minutes and probably merge it if I don't find anything.

**Comment by FishingHacks** [comment]
added a commit

**Comment by RedCrafter07** [comment]
For @git/fetch, will it be pulled? I think it would fail because of the config file... But I'm not sure.

**Comment by FishingHacks** [comment]
Yes, it will be pulled. Why should it fail?

**Review by RedCrafter07 — CHANGES_REQUESTED** [review]
It can't get the remote: (added code for logging the error)
Checking git...
error: No such remote 'origin'

This is a mistake as freshly initialized repositories don't have an origin remote.

**Comment by FishingHacks** [comment]
See the problem, fixed in the next commit

**Comment by RedCrafter07** [comment]
And I'm not really sure about the cwd logic... The cwd should be grabbed from process.cwd() in my opinion

**Comment by FishingHacks** [comment]
> And I'm not really sure about the cwd logic... The cwd should be grabbed from process.cwd() in my opinion

process.cwd() might not reflect the directory, the .rsproj-file is in. However, when someone puts a relative path into the .rsproj-file, they assume it will be relative to the directory, the .rsproj-file is in, not some arbitrary other.

**Comment by FishingHacks** [comment]
> [Fixed @git/fetch, added usage, added contribute.md](https://github.com/RedCrafter07/redstart/pull/3/commits/f8ccef31ae80a8711989e8dd529d26312bd1b7a7)

**Comment by FishingHacks** [comment]
Now it should be fixed

**Comment by RedCrafter07** [comment]
Maybe you have a configs/starter.rsproj file 🤔 
Have you implemented the option to override it in the config?

**Comment by FishingHacks** [comment]
what

**Comment by RedCrafter07** [comment]
Before your changes, there was a workDir: option. Is it implemented here too?

**Comment by FishingHacks** [comment]
no, because top-level (general) config doesn't work yet

**Comment by RedCrafter07** [comment]
What do you mean with "top-level config"?

**Review by RedCrafter07 — CHANGES_REQUESTED** [review]
@git/fetch
fatal: protocol '"https' is not supported

**Comment by FishingHacks** [comment]
smth like this:

```
...modules

workDir: ./ # Doesn't work
dbgprint: true # Doesn't work
[modulename]
key: value # does work
```

**Comment by FishingHacks** [comment]
> @git/fetch fatal: protocol '"https' is not supported

> repository: "https://github.com/..."

is wrong, use

> repository: https://github.com/...

**Comment by RedCrafter07** [comment]
> > @git/fetch fatal: protocol '"https' is not supported
> 
> > repository: "https://github.com/..."
> 
> is wrong, use
> 
> > repository: https://github.com/...

I do so.

**Comment by RedCrafter07** [comment]
```rsproj
echo, @git/fetch

[echo]
message: Hello, World! This should be followed by nothing# When you see this, there's a problem going on :<
color: green

[@git/fetch]
repository: https://github.com/RedCrafter07/redstart.git
```

**Comment by RedCrafter07** [comment]
It's a parsing issue I addressed [here](https://github.com/RedCrafter07/redstart/commit/fe4cf7aba179c7dae3c671b43676628f1f372bd2)

**Comment by FishingHacks** [comment]
```ts
function parseLine(line: string): null | { key: string; value: string } {
    if (line.includes("\n")) throw new Error("line is not a line");
    if (line.match(/^([^:]+): *[^\n\r]+$/)) {
        const split = line.split(":");
        const key = split.shift()?.trim();
        if (!key) return null;
        if (split.length < 1) return null;
        split[0] = split[0].trimStart();
        split[split.length - 1] = split[split.length - 1].trimEnd();
        return { key, value: split.join(":") };
    } else return null;
}
```

**Comment by RedCrafter07** [comment]
It doesn't work tho

**Comment by RedCrafter07** [comment]
> My configuration:
> 
> ```
> echo, @git/fetch
> 
> [echo]
> message: Hello, World! This should be followed by nothing# When you see this, there's a problem going on :<
> color: green
> 
> [@git/fetch]
> repository: https://github.com/RedCrafter07/redstart.git
> ```

Like seen here.

**Comment by FishingHacks** [comment]
To test if it actually works, add

[echo]

message: This is a test :::

, it should print out This is a test :::, and it does that for me

**Comment by RedCrafter07** [comment]
> 

Works for me.

**Comment by FishingHacks** [comment]
`git remote get-url origin`, output?

**Comment by RedCrafter07** [comment]
"https://github.com/RedCrafter07/redstart.git" although defined as seen in my configuration.

File: src/modules/@git/fetch.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                        spawnSync(
+                            "git",
+                            [
+                                "remote",
+                                "add",
+                                "origin",
+                                `"${config.repository}"`,
+                            ],
+                            { cwd }
+                        ).status !== 0
```
Comment (RedCrafter07) [inline src/modules/@git/fetch.ts:42]: This is the mistake.

Comment (FishingHacks) [reply src/modules/@git/fetch.ts:42]: oh, it is

Comment (RedCrafter07) [reply src/modules/@git/fetch.ts:42]: You see, not my mistake 😁

Comment (FishingHacks) [reply src/modules/@git/fetch.ts:42]: I added that, cuz i forgor that it isn't executed like in the shell

Comment (RedCrafter07) [reply src/modules/@git/fetch.ts:42]: Good that I reviewed it in detail 😅

**Comment by FishingHacks** [comment]
should work now

**Comment by RedCrafter07** [comment]
You may use a [prettier](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) like I do so you can help me with formatting in the future 👀 😅

**Comment by RedCrafter07** [comment]
Another error occurred while pulling the repo:
```
Checking git... From https://github.com/RedCrafter07/redstart
 * [new branch]      main       -> origin/main
 * [new tag]         v1.0.1     -> v1.0.1
 * [new tag]         v1.0.0     -> v1.0.0
There is no tracking information for the current branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

If you wish to set tracking information for this branch you can do so with:

    git branch --set-upstream-to=origin/<branch> master


✖ Couldn't fetch remote repository`
```

File: src/modules/@git/fetch.ts [hunk 2]
```diff
@@ -52,7 +52,7 @@ export default {
                             ") already exists",
                     });
             }
-            if (spawnSync("git", ["pull"], { cwd }).status !== 0)
+            if (spawnSync("git", ["pull", "--set-upstream-to=origin/" + (config.branch || "master")], { cwd }).status !== 0)
```
Comment (RedCrafter07) [inline src/modules/@git/fetch.ts:55]: When executing:
```
error: unknown option `set-upstream-to=origin/main'
usage: git pull [<options>] [<repository> [<refspec>...]]
...
```

**Comment by FishingHacks** [comment]
fixed

**Comment by RedCrafter07** [comment]
[!] Branch main not found!

**Comment by RedCrafter07** [comment]
I'm still getting the error 😞

**Comment by FishingHacks** [comment]
ik

**Comment by FishingHacks** [comment]
btw, you should use yaml as the highlighter for .rsproj files. only the @ in [@a/b] is not hightlighted

**Review by RedCrafter07 — CHANGES_REQUESTED** [review]
Branches on remote are failing (added for clearance)

**Comment by RedCrafter07** [comment]
> btw, you should use yaml as the highlighter for .rsproj files. only the @ in [@a/b] is not hightlighted

Oh, true

**Comment by FishingHacks** [comment]
bruh, now it doesn't even add the remote

**Comment by FishingHacks** [comment]
häääää

**Comment by FishingHacks** [comment]
console.log("'git checkout " + config.branch + "'");
spawnSync("git", ["checkout", config.branch])

when i run the command, that gets logged to the console, everything works correctly

**Comment by FishingHacks** [comment]
y

**Comment by FishingHacks** [comment]
lol

**Comment by RedCrafter07** [comment]
I'm sure you'll figure it out 👏

**Comment by RedCrafter07** [comment]
Errors with "Couldn't fetch remote repository"

**Comment by FishingHacks** [comment]
> [Added @test/js, @git/fetch now really works](https://github.com/RedCrafter07/redstart/pull/3/commits/d52a47c2b9639340f67e0ca2e014aca6dffbe78a)

**Review by FishingHacks — COMMENTED** [review]
Everything works as expected!

**Comment by RedCrafter07** [comment]
It works. Nice

**Comment by RedCrafter07** [comment]
But you logged a git command

**Review by RedCrafter07 — APPROVED** [review]

**Comment by RedCrafter07** [comment]
Fixed.

**Comment by FishingHacks** [comment]
thx

**Pull request merged**